File: DIP/Solution/MarsRoverKata/MissionManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using MarsRoverKata.Behaviors;
using MarsRoverKata.Commands;

namespace MarsRoverKata
{
    public class MissionManager
    {
        public Rover Rover { get; private set; }
        public Mars Planet { get; private set; }
        private readonly Commander _commander;
        
        public MissionManager(Rover rover)
        {
            Rover = rover;
            Planet = rover.Mars;
            _commander = new Commander();
            _commander.CommandExecuted += UpdateAliens;
        }

        private void UpdateAliens(object sender, EventArgs e)
        {
            Planet.UpdateAliens();
        }

        public string AcceptCommands(String commandString)
        {
            _commander.Commands = new List<ICommand>();

            var commands = commandString.ToLower().ToCharArray();
            bool success = true;
            foreach (Char command in commands)
            {
                success = AcceptCommand(success, command);
            }

            return success ? String.Empty : "Some invalid commands were found!!!";
        }

        private bool AcceptCommand(bool success, Char command)
        {
            if (command.Equals('f'))
                _commander.Accept(new ForwardCommand(Rover));
            else if (command.Equals('b'))
                _commander.Accept(new BackwardCommand(Rover));
            else if (command.Equals('r'))
                _commander.Accept(new TurnRightCommand(Rover));
            else if (command.Equals('l'))
                _commander.Accept(new TurnLeftCommand(Rover));
            else if (command.Equals('m'))
                _commander.Accept(new FireMissileCommand(Rover));
            else if (command.Equals('g'))
                _commander.Accept(new FireMortarCommand(Rover));
            else
                success = false;
            return success;
        }

        public string ExecuteMission()
        {
            bool success = _commander.ExecuteCommands();

            return success ? String.Empty : "An error occured while executing commands";
        }

        public void AddObstacle(int x, int y, string type, string behavior)
        {
            Point location = new Point(x, y);
            var obstacle = CreateObstacle(location, type, behavior);
            Planet.AddObstacle(obstacle);
        }

        private IObstacle CreateObstacle(Point location, string type, string behavior)
        {
            if (type.Equals("Alien", StringComparison.OrdinalIgnoreCase))
            {
                if (behavior.Equals("tracker", StringComparison.OrdinalIgnoreCase))
                {
                    return new Alien(Planet, location, new Tracker(Rover));
                }
                if (behavior.Equals("wallbuilder", StringComparison.OrdinalIgnoreCase))
                {
                    return new Alien(Planet, location, new WallBuilder(Planet));
                }
                if (behavior.Equals("shooter", StringComparison.OrdinalIgnoreCase))
                {
                    return new Alien(Planet, location, new DoNothing());
                }

                return new Alien(Planet, location, new DoNothing());
            }
            return new Obstacle(location);
        }
    }
}

File: DIP/Java_Final/MarsRoverKata/src/main/java/MarsRoverKata/MissionManager.java
package MarsRoverKata;

import MarsRoverKata.Behaviors.DoNothing;
import MarsRoverKata.Behaviors.Shooter;
import MarsRoverKata.Behaviors.Tracker;
import MarsRoverKata.Behaviors.WallBuilder;
import MarsRoverKata.Commands.*;
import org.apache.commons.lang3.StringUtils;

import java.beans.EventHandler;
import java.util.ArrayList;

public class MissionManager {
    private Rover rover;

    public Rover getRover() {
        return rover;
    }

    public Mars planet;

    public Mars getPlanet() {
        return planet;
    }

    private final Commander _commander;

    public MissionManager(Rover rover) {
        this.rover = rover;
        planet = rover.getMars();
        _commander = new Commander();
        EventHandler eventHandler = new EventHandler(this, "updateAliens", null, null);
        _commander.setEventHandler(eventHandler);
    }

    public void updateAliens() {
        getPlanet().updateAliens();
    }

    public String acceptCommands(String commandString) {
        _commander.setCommands(new ArrayList<ICommand>());

        char[] commands = commandString.toLowerCase().toCharArray();
        boolean success = true;
        for (char command : commands) {
            success = acceptCommand(success, command);
        }

        return success ? StringUtils.EMPTY : "Some invalid commands were found!!!";
    }

    private boolean acceptCommand(boolean success, char command) {
        if (command == 'f')
            _commander.accept(new ForwardCommand(rover));
        else if (command == 'b')
            _commander.accept(new BackwardCommand(rover));
        else if (command == 'r')
            _commander.accept(new TurnRightCommand(rover));
        else if (command == 'l')
            _commander.accept(new TurnLeftCommand(rover));
        else if (command == 'm')
            _commander.accept(new FireMissileCommand(rover));
        else if (command == 'g')
            _commander.accept(new FireMortarCommand(rover));
        else
            success = false;
        return success;
    }

    public String executeMission() {
        boolean success = _commander.executeCommands();
        return success ? StringUtils.EMPTY : "An error occured while executing commands";
    }

    public void addObstacle(int x, int y, String type, String behavior) {
        Point location = new Point(x, y);
        IObstacle obstacle = createObstacle(location, type, behavior);
        getPlanet().addObstacle(obstacle);
    }

    private IObstacle createObstacle(Point location, String type, String behavior) {
        if (type.equalsIgnoreCase("Alien")) {
            if (behavior.equalsIgnoreCase("tracker"))
            {
                return new Alien(getPlanet(), location, new Tracker(getRover()));
            }
            if (behavior.equalsIgnoreCase("wallbuilder"))
            {
                return new Alien(getPlanet(), location, new WallBuilder(getPlanet()));
            }
            if (behavior.equalsIgnoreCase("shooter"))
            {
                return new Alien(getPlanet(), location, new Shooter());
            }

            return new Alien(getPlanet(), location, new DoNothing());
        }
        return new Obstacle(location);
    }
}

